File: csharp/3_-_Longest_Substring_Without_Repeating_Characters.cs
/**
 * 3. Longest Substring Without Repeating Characters - Medium
 * Given a string s, find the length of the longest substring without repeating characters.
 */

public class Solution {
    public int LengthOfLongestSubstring(string s) {
        Dictionary<char,int> freq = new Dictionary<char,int>();
        int ans = 0;
        
        int begin = 0;
        int end = 0;
        
        while (end < s.Length) {
            char endChar = s[end];
            
            if (!freq.ContainsKey(endChar)) {
                freq[endChar] = 1;
                ans = Math.Max(ans, freq.Count);
                ++end;
            } else {
                char beginChar = s[begin];
                freq.Remove(beginChar);
                ++begin;
            }
        }
        
        return ans;
    }
}

File: javascript/3_-_Longest_Substring_Without_Repeating_Characters.js
/**
 * 3. Longest Substring Without Repeating Characters - Medium
 * Given a string s, find the length of the longest substring without repeating characters.
 */

/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    const freq = new Map();
    let ans = 0;

    let begin = 0;
    let end = 0;

    while (end < s.length) {
        const endChar = s.charAt(end);

        if (!freq.has(endChar)) {
            freq.set(endChar, 1);
            ans = Math.max(ans, freq.size);
            ++end;
        } else {
            const beginChar = s.charAt(begin);
            freq.delete(beginChar);
            ++begin;
        }
    }

    return ans;
};